File: exercises/D307/solution.pas
var
    s: array[1..3] of string;
    i, j: longint;
    c: char;
begin
    readln(s[1]);
    readln(s[2]);
    readln(s[3]);
    for i := 1 to 2 do
    begin
        if (i = 1) then
            c := 'O'
        else
            c := 'X';
        for j := 1 to 3 do
        begin
            if ((s[j][1] = c) and (s[j][2] = c) and (s[j][3] = c)) then
            begin
                writeln(c, ' wins');
                halt
            end;
            if ((s[1][j] = c) and (s[2][j] = c) and (s[3][j] = c)) then
            begin
                writeln(c, ' wins');
                halt
            end
        end;
        if (((s[1][1] = c) and (s[2][2] = c) and (s[3][3] = c)) or
            ((s[1][3] = c) and (s[2][2] = c) and (s[3][1] = c))) then
            begin
                writeln(c, ' wins');
                halt
            end            
    end;
    for i := 1 to 3 do
        for j := 1 to 3 do
            if (s[i][j] = '.') then
            begin
                writeln('Not ended');
                halt
            end;
    writeln('Draw')
end.
        

File: exercises/D307/solution.py
import sys

s = []
for i in range(3):
    line = input()
    s.append(line)
for i in range(2):
    c = 'O'
    if i == 1:
        c = 'X'
    for j in range(3):
        if s[j][0] == c and s[j][1] == c and s[j][2] == c:
            print(c + ' wins')
            sys.exit(0)
        if s[0][j] == c and s[1][j] == c and s[2][j] == c:
            print(c + ' wins')
            sys.exit(0)
    if s[0][0] == c and s[1][1] == c and s[2][2] == c or \
            s[0][2] == c and s[1][1] == c and s[2][0] == c:
        print(c + ' wins')
        sys.exit(0)
for i in range(3):
    for j in range(3):
        if s[i][j] == '.':
            print('Not ended')
            sys.exit(0)
print('Draw')
